### ISSUE TITLE
VK_initialization_error

### ISSUE THREAD

**Issue by enthalpie** [felixdoerre/primus_vk#15]
Hello, I am no programer but tried to set up primus_vk. I run Ubuntu 18.04.1 on an intel-7700HQ with a Geforce 1060 GTX and bumblebee enabled and working. All attempts to start vulkan-based applications gave me the swpchain error so far, so I tried with primus_vk. The only app that works now - sometimes - is the native cube demo in vk. Sometimes it crashes tho with an initialization deadlock.

The other applications tho, give me the error VK_INITIALIZATION_ERROR. What I tried are the cube demo via wine-staging 3.2, DOTA 2 in Linux.
Trieing to run the applications on the intel CPU works like a charm.

Is this the deadlock error and is there anything I can do about it?
If you need error reports, let me know which, I am on my work PC right now but can post them later.

Thx and greetings,
~ent

**Comment by enthalpie** [comment]
vulkan-cube sometimes work.
Output and Screenshot:

[cube_output.txt](https://github.com/felixdoerre/primus_vk/files/2596961/cube_output.txt)
![bildschirmfoto vom 2018-11-19 21-30-57](https://user-images.githubusercontent.com/3176329/48733314-9402ac80-ec42-11e8-8c51-947673d14874.png)

**Comment by PierfrancescoArdino** [comment]
Can you post your nvidia_icd.json and primus_vk.json?

**Comment by felixdoerre** [comment]
The `/build/vulkan-tools-WIuZIg/vulkan-tools-1.1.82.0+dfsg1/vulkaninfo/vulkaninfo.c:3618: failed with VK_ERROR_INITIALIZATION_FAILED` is caused by a missing emulation of the `EnumeratePhysicalDeviceGroupsKHR` function. That should be implemented. It would be interesting to know if this is also what makes your game fail, and thereby know how important that function is. 
To check for this you could run your game with `gdb --args` (something like `gdb --args wine dota.exe`) and set a breakpoint at `vkEnumeratePhysicalDeviceGroupsKHR` (with `b vkEnumeratePhysicalDeviceGroupsKHR`) before starting the application (with `run`). If the breakpoint gets hit it is very likely that the game has exactly this issue and would profit form a fix.

I don't see this error on my computer in vulkaninfo as my vulkan-sdk is older. So I hope it's just that DOTA 2 uses `vkEnumeratePhysicalDeviceGroupsKHR` and that an easy fix there could help that game. Maybe I've got some time later this week to provide you a draft version to test with DOTA 2.

About the deadlock you probably cannot do anything (except perhaps tedious reversing/breakpointing with gdb to force your application over the deadlock). I am working with the maintainers of the vulkan-loader to resolve this issue properly but that takes time.

Of course I can post my nvidia_icd.json and primus_vk.json, however from the vulkan-debug output you provided they seem fine:

nvidia_icd.json:
```
{
    "file_format_version" : "1.0.0",
    "ICD": {
        "library_path": "/path/to/primus-vk/libnv_vulkan_wrapper.so",
        "api_version" : "1.0.65"
    }
}
```
primus_vk.json:
```
{
  "file_format_version" : "1.0.0",
  "layer" : {
    "name": "VK_LAYER_PRIMUS_PrimusVK",
    "type": "GLOBAL",
    "library_path": "/path/to/primus-vk/libprimus_vk.so",
    "api_version": "1.0.0",
    "implementation_version": "1",
    "description": "Primus-vk - https://github.com/felixdoerre/primus_vk",
    "functions": {
      "vkGetInstanceProcAddr": "PrimusVK_GetInstanceProcAddr",
      "vkGetDeviceProcAddr": "PrimusVK_GetDeviceProcAddr"
    },
    "enable_environment": {
      "ENABLE_PRIMUS_LAYER": "1"
    },
    "disable_environment": {
      "DISABLE_PRIMUS_LAYER": "1"
    }
  }
}
```

**Comment by enthalpie** [comment]
Hi Felix,

all good you are doing fine, I am glad someone is adressing those issues at all. :)

I tried to launch dota with the gdb --args command but it falls back to my intel graphics. It launches with vulkan drivers and runs well but on intel. BBSwitch is saying off and you can hear it from the gpu fans noise, which is none. I think this command is to much encapsulated links: 

gdb --args env PRIMUS_VK_MULTITHREADING=1 ENABLE_PRIMUS_LAYER=1 optirun -b primus steam steam://rungameid/570

Using gdb in Steam launch options is of no use, ofc. 

The main thing I am working on isnt DOTA tho, I could play that with OpenGL drivers and all good or with vulkan on my Intel graphics. I wanted to test Fallout76 with dxvk drivers. According to people on reddit it works, I cant get it to run tho on a hybrid setup. So I wanted to check if dxvk is working at all.

But I think it is all good with dxvk and primus_vk the problem is elsewhere. When I look at the error log, the deadlock gets skipped, like you suspected with real applications. If you have the time and maybe could take a look, I am glad for any input; but the issue doesnt seem to be your code !


[error_log.txt](https://github.com/felixdoerre/primus_vk/files/2609323/error_log.txt)

**Comment by felixdoerre** [comment]
It looks like it might be primus_vk:

Is seems to crash here: https://github.com/felixdoerre/primus_vk/blob/c29caf3ec710ed231032ab33c382fcb2502af444/primus_vk.cpp#L697
That most likely calls https://github.com/KhronosGroup/Vulkan-Loader/blob/master/loader/wsi.c#L393

Do you have debug symbols for libvulkan.so.1 from your operating system? That would make the backtrace more expressive:

```
Backtrace:
=>0 0x00007f9f1695ec48 in libvulkan.so.1 (+0x1fc48) (0x0000000063dfc4d0)
  1 0x00007f9f1696a69e in libvulkan.so.1 (+0x2b69d) (0x0000000063dfc4d0)
  2 0x00007f9cdfec9f20 PrimusVK_CreateSwapchainKHR+0x52d() in libprimus_vk.so (0x0000000063dfc5c0)
  3 0x00007f9f27362582 in winex11 (+0x32581) (0x000000007db21450)
  4 0x00007f9f29409700 wine_vkCreateSwapchainKHR+0xaf() in winevulkan (0x0000000063dfc750)
  5 0x000000006a3d5509 in d3d11 (+0x95508) (0x000000000000002c)
  6 0x000000006a3d5bc0 in d3d11 (+0x95bbf) (0x00000000060aa238)
  7 0x000000006a3b4a2f in d3d11 (+0x74a2e) (0x00000000060aa238)
  8 0x000000006a372175 in d3d11 (+0x32174) (0x0000000000000001)
  9 0x000000006a37266a in d3d11 (+0x32669) (0x0000000000000001)
  10 0x000000006f209585 in dxgi (+0x9584) (0x0000000000000001)
  11 0x000000014238bc72 in fallout76 (+0x238bc71) (0x0000000063dfffd0)
  12 0x000000014103ad39 in fallout76 (+0x103ad38) (0x0000000063dfffd0)
  13 0x000000014103b2b5 in fallout76 (+0x103b2b4) (0x0000000063dfffd0)
  14 0x0000000141bf0989 in fallout76 (+0x1bf0988) (0x0000000063dfffd0)
  15 0x000000007bcb3303 DbgUserBreakPoint+0xaca() in ntdll (0x0000000063dfffd0)
```

**Comment by enthalpie** [comment]
Thanks for your ideas but I fucked up. I reverted to prime.

:/

**Comment by felixdoerre** [comment]
As Dota 2 for linux is free to play, I installed it and tried/debugged it. On my system it wasn't a crash but the game hangs refusing to display any image. The cause is that Dota 2 requires 4 swapchain images while primus_vk always enforced 3. Now the image count is only capped by 3 and when the application requests more, primus_vk goes with that. The fix is in 55fd09279ca169f2cd307d00b2ee708b98850f57. With that change, dota 2 runs flawlessly on my system.

**Comment by felixdoerre** [comment]
I've just implemented `EnumeratePhysicalDeviceGroupsKHR` so this version of vulkaninfo should not report an error anymore. Therefore I consider all issues mentioned here addressed.
